src/accountsdb/cache.zig

```diff
@@ -0,0 +1,209 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const builtin = @import("builtin");
+const zstd = @import("zstd");
+
+const Account = sig.core.Account;
+const Pubkey = sig.core.pubkey.Pubkey;
+const LruCacheCustom = sig.common.lru.LruCacheCustom;
+const ReferenceCounter = sig.sync.reference_counter.ReferenceCounter;
+
+/// Stores read-only in-memory copies of commonly used *rooted* accounts
+pub const AccountsCache = struct {
+    lru: LRU,
+    allocator: std.mem.Allocator,
+
+    /// Atomically refcounted account
+    pub const CachedAccount = struct {
+        account: Account,
+        ref_count: ReferenceCounter,
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) error{OutOfMemory}!CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = .{},
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            if (self.ref_count.release()) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            }
+        }
+
+        // satisfies the type LruCacheCustom expects
+        pub fn releaseOrDestroyDoublePtr(self: **CachedAccount, allocator: std.mem.Allocator) void {
```

Comment:
This is a hack to fix something that we ourselves wrote. `LruCacheCustom` should be patched so as not to need this.

Reply:
Might be a complicated patch to retain compatibility with the situation where you're storing values instead of pointers in the cache. If your other suggestion is implemented, I don't think we need this function any more

> This allocation seems superfluous to me. The LRU already allocates new keys when they are added. Can we rewrite something somewhere to not need this second allocation on every put instead of on the first one?

Reply:
Simplified with https://github.com/Syndica/sig/pull/297/commits/4ad2cb1e2a5f853c1041c03d46dc013536737e76